File: c/triangle/triangle.c
#include "triangle.h"

#define TRUE 1
#define FALSE 0

int is_valid_triangle(triangle_t triangle);
int all_positive_sides(triangle_t triangle);
int satisfy_inequalities(triangle_t triangle);

int is_scalene(triangle_t triangle) {
  if (!is_valid_triangle(triangle)) {
    return FALSE;
  }
  const double a = triangle.a, b = triangle.b, c = triangle.c;
  return a != b && a != c && b != c;
}

int is_isosceles(triangle_t triangle) {
  if (!is_valid_triangle(triangle)) {
    return FALSE;
  }
  const double a = triangle.a, b = triangle.b, c = triangle.c;
  return a == b || a == c || b == c;
}

int is_equilateral(triangle_t triangle) {
  if (!is_valid_triangle(triangle)) {
    return FALSE;
  }
  const double a = triangle.a, b = triangle.b, c = triangle.c;
  return a == b && b == c;
}

int is_valid_triangle(triangle_t triangle) {
  return all_positive_sides(triangle) && satisfy_inequalities(triangle);
}

int all_positive_sides(triangle_t triangle) {
  return triangle.a > 0 && triangle.b > 0 && triangle.c > 0;
}

int satisfy_inequalities(triangle_t triangle) {
  const double a = triangle.a, b = triangle.b, c = triangle.c;
  return a + b >= c && a + c >= b && b + c >= a;
}

File: typescript/triangle/triangle.ts
export class Triangle {
  public static satisfyInequalities(a: number, b: number, c: number): boolean {
    return [
      [a, b, c],
      [a, c, b],
      [b, c, a],
    ].every(([a, b, c]) => a + b >= c);
  }

  private _a: number;
  private _b: number;
  private _c: number;
  private isTriangle: boolean;

  constructor(a: number, b: number, c: number) {
    [this._a, this._b, this._c] = [a, b, c];
    this.isTriangle =
      [a, b, c].every((x) => x > 0) && Triangle.satisfyInequalities(a, b, c);
  }

  get isEquilateral() {
    return this.isTriangle && this._a === this._b && this._b === this._c;
  }

  get isIsosceles() {
    return (
      this.isTriangle &&
      (this._a === this._b || this._a === this._c || this._b === this._c)
    );
  }

  get isScalene() {
    return (
      this.isTriangle &&
      this._a !== this._b &&
      this._a !== this._c &&
      this._b !== this._c
    );
  }
}
